homeassistant/components/bmw_connected_drive/number.py

```diff
@@ -0,0 +1,115 @@
+    remote_service: Callable[[MyBMWVehicle, float | int], Coroutine[Any, Any, Any]]
+
+
+@dataclass
+class BMWNumberEntityDescription(NumberEntityDescription, BMWRequiredKeysMixin):
+    """Describes BMW number entity."""
+
+    is_available: Callable[[MyBMWVehicle], bool] = lambda _: False
+    dynamic_options: Callable[[MyBMWVehicle], list[str]] | None = None
+    mode: NumberMode = NumberMode.AUTO
+
+
+NUMBER_TYPES: list[BMWNumberEntityDescription] = [
+    BMWNumberEntityDescription(
+        key="target_soc",
+        name="Target SoC",
+        device_class=NumberDeviceClass.BATTERY,
+        is_available=lambda v: v.is_remote_set_target_soc_enabled,
+        native_max_value=100.0,
+        native_min_value=20.0,
+        native_step=5.0,
+        mode=NumberMode.SLIDER,
+        value_fn=lambda v: v.fuel_and_battery.charging_target,
+        remote_service=lambda v, o: v.remote_services.trigger_charging_settings_update(
+            target_soc=int(o)
+        ),
+        icon="mdi:battery-charging-medium",
+    ),
+]
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the MyBMW number from config entry."""
+    coordinator: BMWDataUpdateCoordinator = hass.data[DOMAIN][config_entry.entry_id]
+
+    entities: list[BMWNumber] = []
+
+    for vehicle in coordinator.account.vehicles:
+        if not coordinator.read_only:
+            entities.extend(
+                [
+                    BMWNumber(coordinator, vehicle, description)
+                    for description in NUMBER_TYPES
+                    if description.is_available(vehicle)
+                ]
+            )
+    async_add_entities(entities)
+
+
+class BMWNumber(BMWBaseEntity, NumberEntity):
+    """Representation of BMW Number entity."""
+
+    entity_description: BMWNumberEntityDescription
+
+    def __init__(
+        self,
+        coordinator: BMWDataUpdateCoordinator,
+        vehicle: MyBMWVehicle,
+        description: BMWNumberEntityDescription,
+    ) -> None:
+        """Initialize an BMW Number."""
+        super().__init__(coordinator, vehicle)
+        self.entity_description = description
+        self._attr_unique_id = f"{vehicle.vin}-{description.key}"
+        self._attr_native_value = description.value_fn(vehicle)
+        self._attr_mode = description.mode
+
+    @property
+    def native_value(self) -> float | None:
+        """Return the entity value to represent the entity state."""
+        return self.entity_description.value_fn(self.vehicle)
+
+    async def async_set_native_value(self, value: float) -> None:
+        """Update to the vehicle."""
+        _LOGGER.debug(
+            "Executing '%s' on vehicle '%s' to value '%s'",
+            self.entity_description.key,
+            self.vehicle.vin,
+            value,
+        )
+        await self.entity_description.remote_service(self.vehicle, value)
```

Comment:
What is this call fails? Can it raise? should it be handled?

Reply:
The library could raise a `ValueError` in case the given number is not a multiple of 5. 
If communication with the BMW APIs fail, multiple errors could be raised: `httpx.HTTPStatusError`, `TimeoutError` (if the API cannot reach the vehicle in time) or a plain exception.

In my mind, we don't need additional error handling in here, as each failure mode is an actual failure we cannot recover from (well, we could retry but that usually doesn't help much - either the API works and it runs through, or there is something wrong and we have to wait for the API/car to recover).

Reply:
We do need to handle them. Only ValueError and HomeAssistantError is handled upstream. The rest are bugs and have side effects.

Reply:
But I suppose a broad `except Exception as ex` would not be the way to go? :)

I could implement a custom error class in the library (or rather reuse the ones implemented for #90274 and add more for the remote services in an additional libary version). 

However I am unsure how to best continue with this: Adding error handling should then be done in this PR, added to `select` from #88759 and implemented for [bmw_connected_drive/button.py#L129](https://github.com/home-assistant/core/blob/dev/homeassistant/components/bmw_connected_drive/button.py#L129).

As I won't be able to get something else than `except Exception as ex` through until the beta cut, please advise if how you would like to continue:
- broad exception
- keep as is and I'll create another PR in the next days that fixes exception handling for all above (should be able to fix it during beta fix if required)
- don't merge this (and then have a breaking change for 2023.6)

Reply:
Hope this satisfies the requirement (and maybe it can be squeezed into this beta?). Otherwise this will be breaking change for 2022.6 and I'll add the breaking change paragraph again.